In [1]:
from astropy.table import Table, join, Column
import matplotlib.pyplot as plt
import numpy as np
# My modules
from my_progs.catalog.vsh_deg1_cor import vsh_deg01_fitting
from my_progs.catalog.vsh_deg2_cor import vsh_deg02_fitting
from my_progs.catalog.pos_diff import pos_diff_calc
from my_progs.catalog.pos_err import pos_err_calc

Read these two catalogs.

In [2]:
# read source position from VLBI and Gaia catalogs.
from my_progs.vlbi.read_sou import read_cat
tv = read_cat("/Users/Neo/Astronomy/Data/VLBISolutions/"
                "vlbi2_server/GalacticAberration/GaiaDR2-timespan-solutions/"
                "gaiadr2-timespan-noGAb.cat")

# Read Gaia DR2 IERS quasar
# tg = Table.read("/Users/Neo/Astronomy/Data/catalogs/Gaia_DR2/"
#                 "gaiadr2_iers.fits")
from my_progs.catalog.read_gaia import read_gaiadr2_qso

tg = read_gaiadr2_qso()

Then I obtained a sample of common sources by the cross-match.

In [4]:
# Cross-match
t_com = join(tv, tg, keys="iers_name", table_names=["v", "g"])

# gaiadr2-vlbi positional offsets (VLBI - Gaia DR2).
[dra, ddec, dra_err, ddec_err, dra_ddec_cor,
 ang_sep, X_a, X_d, X, X2] = pos_diff_calc(
    t_com["ra_v"], t_com["ra_err"], t_com["dec_v"], t_com["dec_err"], t_com["ra_dec_corr_v"],
    t_com["ra_g"],  t_com["ra_error"], t_com["dec_g"], t_com["dec_error"], t_com["ra_dec_corr_g"])

dra_ddec_cov = dra_ddec_cor * dra_err * ddec_err

# # degree -> rad
# ra = np.array(t_com["ra_1"])
# dec = np.array(t_com["dec_1"])
# ra_rad, dec_rad = np.deg2rad(ra), np.deg2rad(dec)

t_com.add_columns([dra, ddec, dra_err, ddec_err, dra_ddec_cor,
                   ang_sep, X_a, X_d, X, X2],
                  names=["dra", "ddec", "dra_err", "ddec_err", "dra_ddec_cov",
                         "ang_sep", "nor_dra",  "nor_ddec", "nor_sep_cov", "nor_sep"])

t_com

<Table masked=True length=1274>
ivs_name iers_name      ra_v      ...    nor_sep_cov          nor_sep      
                        deg       ...        deg                deg        
  str8      str8      float64     ...      float64            float64      
-------- --------- -------------- ... ------------------ ------------------
0000-197  0000-197   0.8278125995 ... 1.9871789731405747 2.1180808213411035
0001+459  0001+459   1.0671985141 ... 1.3522894213624823 1.3162036361974871
0002-170  0002-170   1.3247240786 ...  0.793452109707901 0.7992763571166984
0002-478  0002-478   1.1485645299 ... 0.5504634852830819 0.6051432698720541
0003+158  0003+158   1.4968240525 ... 0.6089579314488115 0.5954152377593939
0003+380  0003+380   1.4882308924 ...  8.003304637386364  2.705773221087961
0003-066  0003-066    1.557887033 ... 3.3930340519363527  2.872547950516595
0006+061  0006+061   2.2663826906 ...  9.314191206731586  9.503617218598738
0006+397  0006+397   2.2673901042 ...  2.565833610582421 2.4911708370957197
     ...       ...            ... ...                ...                ...
2350+704  2350+704 358.2202306798 ... 1.1913931776006033  1.164640689206144
2351-154  2351-154 358.6258131756 ... 1.0533531065158064 1.0459169577041005
2353+816  2353+816 359.0949746553 ... 0.9723107014453095 0.8351688921158756
2353-686  2353-686  359.002839252 ...  5.734140586659047  6.151426599280167
2354-117  2354-117 359.3799897768 ... 1.2729469919818766 1.3304297096941906
2355+490  2355+490 359.5408897024 ... 0.8670785996617463 0.8459568567715082
2355-106  2355-106 359.5453433167 ... 1.4767930316045443 1.4725685749293123
2355-534  2355-534 359.4719418747 ...  4.867386291428503  4.519810899308275
2356+385  2356+385 359.8882532855 ... 1.6912534598580926 1.6388989051185785
2357-318  2357-318 359.8978813402 ... 0.7357346973952129 0.7355295974736972

There are 1274 common sources. 
Before we investigate the systematcis, we should remove some outliers.

In [6]:
# First we remove sources with less than three observations.
mask1 = (t_com["num_del"] >=3)
t_com1 = t_com[mask1]
t_com1

<Table masked=True length=1263>
ivs_name iers_name      ra_v      ...    nor_sep_cov          nor_sep      
                        deg       ...        deg                deg        
  str8      str8      float64     ...      float64            float64      
-------- --------- -------------- ... ------------------ ------------------
0000-197  0000-197   0.8278125995 ... 1.9871789731405747 2.1180808213411035
0001+459  0001+459   1.0671985141 ... 1.3522894213624823 1.3162036361974871
0002-170  0002-170   1.3247240786 ...  0.793452109707901 0.7992763571166984
0002-478  0002-478   1.1485645299 ... 0.5504634852830819 0.6051432698720541
0003+158  0003+158   1.4968240525 ... 0.6089579314488115 0.5954152377593939
0003+380  0003+380   1.4882308924 ...  8.003304637386364  2.705773221087961
0003-066  0003-066    1.557887033 ... 3.3930340519363527  2.872547950516595
0006+061  0006+061   2.2663826906 ...  9.314191206731586  9.503617218598738
0006+397  0006+397   2.2673901042 ...  2.565833610582421 2.4911708370957197
     ...       ...            ... ...                ...                ...
2350+704  2350+704 358.2202306798 ... 1.1913931776006033  1.164640689206144
2351-154  2351-154 358.6258131756 ... 1.0533531065158064 1.0459169577041005
2353+816  2353+816 359.0949746553 ... 0.9723107014453095 0.8351688921158756
2353-686  2353-686  359.002839252 ...  5.734140586659047  6.151426599280167
2354-117  2354-117 359.3799897768 ... 1.2729469919818766 1.3304297096941906
2355+490  2355+490 359.5408897024 ... 0.8670785996617463 0.8459568567715082
2355-106  2355-106 359.5453433167 ... 1.4767930316045443 1.4725685749293123
2355-534  2355-534 359.4719418747 ...  4.867386291428503  4.519810899308275
2356+385  2356+385 359.8882532855 ... 1.6912534598580926 1.6388989051185785
2357-318  2357-318 359.8978813402 ... 0.7357346973952129 0.7355295974736972

In [9]:
# Consider X follow a Rayleigh distribution
N = len(t_com1)
X0 = np.sqrt(2 * np.log(N))

print("There are %d sources.\n"
      "When X >= %.2f, the number of the outlier is smaller than 1." % (N, X0))

# Then we remove these sources with X >= 3.8 or eema >= 10 mas
mask2 = ((t_com1["nor_sep_cov"] <= 3.8) & (t_com1["ang_sep"] <= 10) & (
    t_com1["pos_err_v"] <= 10) & (t_com1["pos_err_g"] <= 10))
t_com2 = t_com1[mask2]
t_com2

There are 1263 sources.
When X >= 3.78, the number of the outlier is smaller than 1.


<Table masked=True length=1024>
ivs_name iers_name      ra_v      ...    nor_sep_cov          nor_sep      
                        deg       ...        deg                deg        
  str8      str8      float64     ...      float64            float64      
-------- --------- -------------- ... ------------------ ------------------
0000-197  0000-197   0.8278125995 ... 1.9871789731405747 2.1180808213411035
0001+459  0001+459   1.0671985141 ... 1.3522894213624823 1.3162036361974871
0002-170  0002-170   1.3247240786 ...  0.793452109707901 0.7992763571166984
0002-478  0002-478   1.1485645299 ... 0.5504634852830819 0.6051432698720541
0003+158  0003+158   1.4968240525 ... 0.6089579314488115 0.5954152377593939
0003-066  0003-066    1.557887033 ... 3.3930340519363527  2.872547950516595
0006+397  0006+397   2.2673901042 ...  2.565833610582421 2.4911708370957197
0007+439  0007+439   2.6251936172 ...  1.367369063388493  1.346335794624737
0007-048  0007-048   2.5016194008 ... 1.4100548886308908 1.6443526755670825
     ...       ...            ... ...                ...                ...
2344-514  2344-514 356.8327670123 ... 2.1648335786491715  2.184053806989794
2346-049  2346-049 357.2923592151 ... 0.7817090588172241  0.735635976175877
2350+704  2350+704 358.2202306798 ... 1.1913931776006033  1.164640689206144
2351-154  2351-154 358.6258131756 ... 1.0533531065158064 1.0459169577041005
2353+816  2353+816 359.0949746553 ... 0.9723107014453095 0.8351688921158756
2354-117  2354-117 359.3799897768 ... 1.2729469919818766 1.3304297096941906
2355+490  2355+490 359.5408897024 ... 0.8670785996617463 0.8459568567715082
2355-106  2355-106 359.5453433167 ... 1.4767930316045443 1.4725685749293123
2356+385  2356+385 359.8882532855 ... 1.6912534598580926 1.6388989051185785
2357-318  2357-318 359.8978813402 ... 0.7357346973952129 0.7355295974736972

This left 1024 common source. Actually, if we adopted the X0 to be 3.78 rather than 3.8, there would be 1023 sources left.

In [10]:
import astropy.units as u

# Genearte some arrays for estimating the ststematics
dra = np.array(t_com2["dra"])
ddec = np.array(t_com2["ddec"])
dra_err = np.array(t_com2["dra_err"])
ddec_err = np.array(t_com2["ddec_err"])
dra_ddec_cov = np.array(t_com2["dra_ddec_cov"])
ra_rad = np.array(t_com2["ra_v"].to(u.radian))
dec_rad = np.array(t_com2["dec_v"].to(u.radian))

In [14]:
# l_max = 1
w1, sig1, corrcoef1, _, _, _ = vsh_deg01_fitting(
    dra, ddec, ra_rad, dec_rad, dra_err, ddec_err, 
    cov=dra_ddec_cov, elim_flag="None")

# l_max = 2
w2, sig2, corrcoef2, _, _, _ = vsh_deg02_fitting(
    dra, ddec, ra_rad, dec_rad, dra_err, ddec_err,  
    cov=dra_ddec_cov, elim_flag="None")

# mas -> uas
w1 = w1 * 1.e3
sig1 = sig1 * 1.e3
w2 = w2 * 1.e3
sig2 = sig2 * 1.e3

# Print results
print("Estimates (%6d sources)\n"
      "----------------------------------------------"
      "----------------------------------------------\n"
      "               Rotation [uas]                 "
      "                  Glide [uas]               \n"
      "               x             y             z"
      "               x             y             z\n"
      "----------------------------------------------"
      "----------------------------------------------\n"
      "l_max=1  %+4.0f +/- %3.0f  %+4.0f +/- %3.0f  %+4.0f +/- %3.0f  "
      "  %+4.0f +/- %3.0f  %+4.0f +/- %3.0f  %+4.0f +/- %3.0f\n"
      "l_max=2  %+4.0f +/- %3.0f  %+4.0f +/- %3.0f  %+4.0f +/- %3.0f  "
      "  %+4.0f +/- %3.0f  %+4.0f +/- %3.0f  %+4.0f +/- %3.0f\n"
      "----------------------------------------------"
      "----------------------------------------------\n" %
      (dra.size, w1[3], sig1[3], w1[4], sig1[4], w1[5], sig1[5],
       w1[0], sig1[0], w1[1], sig1[1], w1[2], sig1[2],
       w2[3], sig2[3], w2[4], sig2[4], w2[5], sig2[5],
       w2[0], sig2[0], w2[1], sig2[1], w2[2], sig2[2]))

quad_names = Column(["ER22", "EI22", "ER21", "EI21", "E20",
                     "MR22", "MI22", "MR21", "MI21", "M20"])
t_quad = Table([quad_names, w2[6:], sig2[6:]], names=["Quadrupolar term", "Estimate", "Error"])
t_quad["Estimate"].format = "%5.0f"
t_quad["Error"].format = "%5.0f"
print(t_quad)

# apriori statistics (weighted)
#         mean for RA:      0.021 mas
#         wrms for RA:      0.719 mas
#          std for RA:      0.719 mas
#        mean for Dec:      0.053 mas
#        wrms for Dec:      0.948 mas
#         std for Dec:      0.946 mas

# apriori reduced Chi-square for:      1.797
# posteriori statistics  of vsh01 fit (weighted)
#         mean for RA:      0.014 mas
#          rms for RA:      0.718 mas
#          std for RA:      0.717 mas
#        mean for Dec:      0.064 mas
#         rms for Dec:      0.948 mas
#         std for Dec:      0.946 mas

# posteriori reduced Chi-square for:      1.783
# goodness-of-fit is      0.000
# apriori statistics (weighted)
#         mean for RA:      0.021 mas
#         wrms for RA:      0.719 mas
#          std for RA:      0.719 mas
#        mean for Dec:      0.053 mas
#        wrms for Dec:      0.948 mas
#         std for Dec:      0.946 mas

# apriori reduced Chi-square for:      1.797
# posteriori statistics of vsh

In [ ]:
t_com[t_com["iers_name"] == "1901+319"]